src/main/java/ch/njol/skript/registrations/EventValues.java

```diff
@@ -271,58 +271,79 @@ public T get(E event) {
 				}
 			};
 		}
-		// Fourth check will attempt to convert the event value to the type.
+		// Fourth check will attempt to convert the event value to the requesting type.
+		// This first for loop will check that the events are exact. See issue #5016
 		for (EventValueInfo<?, ?> ev : eventValues) {
-			boolean checkInstanceOf = !ev.event.isAssignableFrom(e);
-			if (checkInstanceOf && !e.isAssignableFrom(ev.event))
+			if (!event.getClass().equals(ev.event))
```

Comment:
`event` is already a `Class<? extends Event>`, calling getClass gives a `Class<Class>` so the equals never returns true.